### DISCUSSION TITLE
XCTUnimplemented with Effects

### DISCUSSION THREAD

**Discussion by Miiha** [pointfreeco/swift-composable-architecture#1199] [General]
Hello,

I recently started adopting the new `XCTUnimplemented` functions provided through [xctest-dynamic-overlay](https://github.com/pointfreeco/xctest-dynamic-overlay). 
It really cleans a lot of boilerplate code for the test setup code. 
I also used it on closures that return an `Effect`, which worked for most of the unit tests.
But for tests that were testing code working with e.g `debounce`, I encountered an odd behaviour and unexpected failures in the tests.

The reason is, that the`.failing` (now `.unimplemented`) effect fails the test when the effect is actually run by the store. 
`XCTUnimplemented` fails immediately when the closure is called, not when the effect is run.

Would it make sense to have a version of `XCTUnimplemented`, dedicated to the TCA's `Effect`, which provides the `.failing`/`.unimplemented` behaviour?
I wonder if others might also assume that  `XCTUnimplemented` would be applicable for scenarios where `Effect.unimlemented` is used.

Cheers,
Micha

**Comment by stephencelis** [comment]
@Miiha This behavior is actually intentional, though nuanced for sure. We think it's perfectly reasonable to delay failure till the dependency is actually used, and allow for effects and clients to be referenced outside of failure. For example:

```swift
let effect = environment.apiClient.fetchCurrentUser()
  .catchToEffect(Action.currentUserResponse)
```

`XCTUnimplemented` is lazy in a similar way, since it is meant to be invoked from inside async contexts like `Effect.task`, which is also delayed.

Does this make sense?

**Reply by Miiha** [reply]
Hi @stephencelis, thanks for clarifying. Now knowing that the library is fully embracing swift concurrency it really makes sense.
